v-next/hardhat/src/internal/builtin-plugins/solidity/build-results.ts

```diff
@@ -54,27 +56,95 @@ export function throwIfSolidityBuildFailed(
   }
 }
 
+/**
+ * This function returns the build infos and outputs associated with the given
+ * Solidity build results.
+ *
+ * @param results The successful Solidity build results.
+ * @param artifactManager The artifact manager.
+ * @returns The build infos in the Hardhat v3 format as expected by the EDR.
+ */
+export async function getBuildInfos(
+  results: SuccessfulSolidityBuildResults,
+  artifactManager: ArtifactManager,
+): Promise<BuildInfoAndOutput[]> {
+  const buildIds = Array.from(new Set(results.values())).map(
+    ({ buildId }) => buildId,
+  );
+
+  return Promise.all(
+    buildIds.map(async (buildId) => {
+      const buildInfoPath = await artifactManager.getBuildInfoPath(buildId);
+      const buildInfoOutputPath =
+        await artifactManager.getBuildInfoOutputPath(buildId);
+
+      assertHardhatInvariant(
+        buildInfoPath !== undefined,
+        "buildInfoPath should not be undefined",
+      );
+      assertHardhatInvariant(
+        buildInfoOutputPath !== undefined,
+        "buildInfoOutputPath should not be undefined",
+      );
+
+      const buildInfo = await readBinaryFile(buildInfoPath);
+      const output = await readBinaryFile(buildInfoOutputPath);
+
+      return {
+        buildInfo,
+        output,
+      };
+    }),
+  );
+}
+
 /**
  * This function returns the artifacts generated during the compilation associated
  * with the given Solidity build results. It relies on the fact that each successful
  * build result has a corresponding artifact generated property.
+ *
+ * @param results The successful Solidity build results.
+ * @param artifactManager The artifact manager.
+ * @returns The artifacts in the format expected by the EDR.
  */
 export async function getArtifacts(
   results: SuccessfulSolidityBuildResults,
-  artifactsRootPath: string,
+  buildInfos: BuildInfoAndOutput[],
 ): Promise<EdrArtifact[]> {
-  const artifacts: EdrArtifact[] = [];
+  const solcVersions = Object.fromEntries(
+    buildInfos.map(({ buildInfo }) => {
+      const solidityBuildInfo: SolidityBuildInfo = JSON.parse(
+        new TextDecoder("utf-8").decode(buildInfo),
+      );
+
+      return [solidityBuildInfo.id, solidityBuildInfo.solcVersion];
+    }),
+  );
 
-  for (const [source, result] of results.entries()) {
-    for (const artifactPath of result.contractArtifactsGenerated) {
+  const contractArtifacts = Array.from(results.entries())
+    .map(([source, result]) => {
+      return result.contractArtifactsGenerated.map((artifactPath) => ({
+        source,
+        buildId: result.buildId,
+        artifactPath,
+      }));
+    })
+    .flat();
+
+  return Promise.all(
+    contractArtifacts.map(async ({ source, buildId, artifactPath }) => {
```

Comment:
I parallelised reading artifacts from disk here in hope that it will speed up the preparation for running solidity tests.